ttnn/cpp/ttnn/operations/data_movement/expand/device/expand_device_operation.cpp

```diff
@@ -0,0 +1,82 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "expand_device_operation.hpp"
+
+#include <cstdint>
+
+#include "ttnn/tensor/tensor.hpp"
+#include "ttnn/tensor/types.hpp"
+
+namespace ttnn::operations::expand {
+ExpandOperation::program_factory_t ExpandOperation::select_program_factory(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    const auto& input = tensor_args.input;
+
+    switch (input.get_layout()) {
+        case Layout::ROW_MAJOR: return ExpandRowMajorFactory{};
+        default: TT_FATAL(false, "Unsupported input layout");
+    }
+}
+
+void validate(
+    const ExpandOperation::operation_attributes_t& operation_attributes,
+    const ExpandOperation::tensor_args_t& tensor_args) {
+    // We need to assert that the input and output are ROW_MAJOR. (unfortunately)
+
+    const auto& input = tensor_args.input;
+    const auto& output = tensor_args.output;
+
+    TT_FATAL(tensor_args.input.storage_type() == StorageType::DEVICE, "Expand: Input tensor need to be on device");
+    TT_FATAL(tensor_args.input.buffer() != nullptr, "Expand: Input tensor need to be allocated in buffers on device");
+
```

Comment:
There are probably more things to check here:
- input layout
- input dtype?
- restriction for the ops in any dims?
- checks on input tensor rank?

Reply:
Here are the current state:
- Layout is already asserted (row-major only)
- dtype should be fine (given that it is doing a byte-copy within the device basically. I think I will need to re-test to make sure it works with bfp8, but fp32 and bf16 I had no issues with for now.
- It should be nd given that the only thing needed for rank > 2 is to duplicate the tensor.
- I think the throw at `infer_size` should handle it already?